### PR TITLE
Major refactor and new features

### PR DESCRIPTION
**Pull request by JZfi**
Note that the ipv6 disabling stuff is completely removed and the kernel numa variable is gone (that stuff can be added to the extra_kernel_params).

### CODE REVIEW

**Comment by khappone** [comment]
Hi Niklas!
Thanks for the big amount of work you've done for this. It looks good, but I have some comments.

First of all if there are incremental changes we agree on, it might be good not to have the PR come from master, so we don't break your master accidentally. Although, if you're OK with that, I'm good :). I'll do the simple stuff in a review, and the more architectural ones probably in comments.

File: templates/node_pxe.conf.j2 [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{% if use_kickstart %}
```
Comment (khappone) [inline templates/node_pxe.conf.j2]: This should probably not be a conditional, it should always be used. I think we have two cases.
1) The DHCP server handles DHCP and kickstart file management (use_kickstart = false)
2) The DHCP server handles DHCP but not kickstart file manangement (use_kickstart = true)

In 2) we still want to set the kickstart URL, but we don't want to generate the configs. Currently it won't give the kickstart parameter if use_kickstart = false.

File: defaults/main.yml [hunk 2]
```diff
@@ -1,3 +1,12 @@
 ---
-dhcp_group: "dhcp_only_hosts"
-dhcp_pxe_group: "dhcp_pxe_hosts"
+lang: "en_US.UTF-8"
```
Comment (khappone) [inline defaults/main.yml]: This is a matter of preference, but generally I prefer to prefix all role-specific variables with a role-specific prefix. I have several times run into issues where variable names are accidentally overloaded, and this generates hard to debug problems. Having a prefix makes it better, even though it doesn't remove the problem.

So, e.g. all variables could be kickstart_. This of course touches a lot of files.

File: tasks/configure.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  when: use_pxe and use_kickstart
+
+- name: Find all the configs in dhcpd.d directory
+  run_once: true
+  find:
+    paths: "/etc/dhcp/dhcpd.d/"
+    patterns: "*.conf"
+  register: configs
+ 
+- name: Clear hosts config
```
Comment (khappone) [inline tasks/configure.yml]: Not a big deal, but this and the next task could be quite easily replaced by a template like this. The main benefits would be a more standard way of doing it, and avoiding unnecessary "changed" tasks, and service restarts.

 - name: Create dhcp include file
    run_once: true
    template: src=dhcp_include.j2 dest="/etc/dhcp/dhcpd_hosts.conf"
    notify: restart dhcpd

**Comment by khappone** [comment]
Ok, then for a a main architectural comment. Your way is prettier than the old way.

However, it uses a pattern that I don't especially like. This is adding role level logic into the calling playbook. Currently the playbook needs quite a bit of logic to call the role. In my opinion the roles would need to stand pretty much on their own.

There could be quite a clean way of solving this. First set a variable, e.g. kickstart_server (can have better name), which is false by default, and true in the scope of the dhcp server.

in tasks/main.yml do conditional includes. Include install.yml only when kickstart_server == true, and configure.yml only when it's false. Then put all tasks in configure.yml inside a block, with a delegate_to: "{{ dhcp_delegate_server }}". This can be specified globally for your nodes.

After this you could just include the roles in the playbook, first for the dhcp server, then for the other nodes without any playbook-level magic.

The only thing this misses is creating the include file, but this can be simply touched in the install.yml.

**Comment by junousi** [comment]
Good refactoring but we can't take it in without the requested changes.

We've rebased it to CSCfi:master and pushed the results to a feature branch there. If at some point we have time to work on the architectural tidbits, we'll do a merge from there. I'll now close this particular PR.

**Comment by oscarkraemer** [comment]
Branch in question: https://github.com/CSCfi/ansible-role-dhcp-kickstart/tree/SoneraCloud_PR_rebase

**Pull request closed without merging**